src/lib/VirtualScroller.js

```diff
@@ -0,0 +1,183 @@
+        if (this.containerEl) {
+            this.containerEl.remove();
+        }
+
+        this.containerEl = null;
+        this.contentEl = null;
+    }
+
+    /**
+     * Initializes the virtual scroller
+     *
+     * @param {Object} config - The config
+     * @return {void}
+     */
+    init(config) {
+        if (!config.totalItems || !config.itemHeight) {
+            throw new Error('Need to provide totalItems and itemHeight');
+        }
+
+        // The total number items to be scrolled
+        this.totalItems = config.totalItems;
+
+        // The height of each individual item
+        this.itemHeight = config.itemHeight;
+
+        // The height of the visible container
+        this.containerHeight = config.containerHeight;
+
+        // The callback function that to allow users generate the item
+        this.renderItemFn = config.renderItemFn;
+
+        // Allows the user to specify the margin at the top of the container before the first item is rendered
+        this.marginTop = config.marginTop || 0;
+
+        // Allows the user to specify the margin at the bottom of the container after the last item is rendered
+        this.marginBottom = config.marginBottom || 0;
+
+        // The number of items that can fit in view
+        this.totalViewItems = Math.ceil(this.containerHeight / this.itemHeight);
+
+        // The height of the buffer before virtual scroll renders the next set
+        this.maxBufferHeight = this.totalViewItems * this.itemHeight;
+
+        // The max number of items to render at any one given time
+        this.maxRenderedItems = Math.ceil(this.totalViewItems * 3);
+
+        // Create the scrolling container element
+        this.containerEl = document.createElement('div');
+        this.containerEl.className = 'vs-container';
+
+        // Create the true height content container
+        this.contentEl = document.createElement('div');
+        this.contentEl.className = 'vs-content-container';
+        this.contentEl.style.height = `${this.totalItems * this.itemHeight + this.marginTop + this.marginBottom}px`;
+
+        this.containerEl.appendChild(this.contentEl);
+        this.container.appendChild(this.containerEl);
+
+        this.renderItems();
+
+        this.bindDOMListeners();
+    }
+
+    /**
+     * Binds DOM listeners
+     *
+     * @return {void}
+     */
+    bindDOMListeners() {
+        this.containerEl.addEventListener('scroll', this.onScrollHandler);
+    }
+
+    /**
+     * Unbinds DOM listeners
+     *
+     * @return {void}
+     */
+    unbindDOMListeners() {
+        if (this.containerEl) {
+            this.containerEl.removeEventListener('scroll', this.onScrollHandler);
+        }
+    }
+
+    /**
+     * Handler for 'scroll' event
+     *
+     * @param {Event} e - The scroll event
+     * @return {void}
+     */
+    onScrollHandler(e) {
+        const { scrollTop } = e.target;
+
+        if (Math.abs(scrollTop - this.previousScrollTop) > this.maxBufferHeight) {
```

Comment:
We're doing a decent amount of math here and scroll events fire extremely often. Do you think the frequency of scroll events could cause a perf problem here?

Reply:
definitely room for improvement, let me address that in a follow up PR